pkg/cache/cache.go

```diff
@@ -187,6 +198,9 @@ type Options struct {
 	// ByObject restricts the cache's ListWatch to the desired fields per GVK at the specified object.
 	// object, this will fall through to Default* settings.
 	ByObject map[client.Object]ByObject
+
+	// NewInformer allows overriding of NewSharedIndexInformer for testing.
```

Comment:
I see two options to avoid exposing this option:

1. Reflection
```go
			v := reflect.ValueOf(informerCache).Elem()
			informersField := v.FieldByName("Informers").Elem()
			newInformerField := informersField.FieldByName("newInformer")
			reflect.NewAt(newInformerField.Type(), unsafe.Pointer(newInformerField.UnsafeAddr())).
				Elem().
				Set(reflect.ValueOf(func(lw kcache.ListerWatcher, o runtime.Object, d time.Duration, i kcache.Indexers) kcache.SharedIndexInformer {
					return &controllertest.FakeInformer{Synced: false}
				}))
```

2. Add an exported func to informerCache to set the NewInformer func. Define an interface with this func in the test, cast to this interface and use the func

Reply:
I personally prefer hacks in unit tests over exporting funcs only for unit tests.

Reply:
Ack, used the reflect hack and moved discussion to the related comment thread below ( https://github.com/kubernetes-sigs/controller-runtime/pull/2371/files/35b504b14c79305b5fee50ea039d514b34c39d62#r1252489000 )

LMK what you think

Reply:
Thx will review in a bit and resolve here